File: python/examples/harness/scenarios/adaptive_search.py
"""Scenarios exercising the adaptive text-search surface end-to-end.

Each sub-scenario opens its own sibling database so the registered FTS
property schemas and seeded nodes do not collide with other harness
scenarios, and teardown falls out of scope when the engine handle is
closed. The scenarios mirror the dedicated fixtures in
``python/tests/test_text_search_surface.py`` and
``crates/fathomdb/tests/text_search_surface.rs`` but drive the public
SDK surface through the scenario runner rather than pytest.
"""

from __future__ import annotations

from fathomdb import (
    ChunkInsert,
    ChunkPolicy,
    FtsPropertyPathMode,
    FtsPropertyPathSpec,
    NodeInsert,
    SearchHitSource,
    SearchMatchMode,
    WriteRequest,
    new_row_id,
)

from ..models import HarnessContext, ScenarioResult


_ADAPTIVE_GOAL_KIND = "AdaptiveGoal"
_ADAPTIVE_KI_KIND = "AdaptiveKnowledgeItem"


def _seed_adaptive_goals(engine) -> None:
    """Seed three Goal-like nodes with deterministic chunk text."""
    engine.write(
        WriteRequest(
            label="adaptive-seed-goals",
            nodes=[
                NodeInsert(
                    row_id=new_row_id(),
                    logical_id="adaptive-goal-budget",
                    kind=_ADAPTIVE_GOAL_KIND,
                    properties={"title": "Budget meeting"},
                    source_ref="adaptive-seed",
                    upsert=False,
                    chunk_policy=ChunkPolicy.PRESERVE,
                ),
                NodeInsert(
                    row_id=new_row_id(),
                    logical_id="adaptive-goal-quarterly",
                    kind=_ADAPTIVE_GOAL_KIND,
                    properties={"title": "Quarterly planning"},
                    source_ref="adaptive-seed",
                    upsert=False,
                    chunk_policy=ChunkPolicy.PRESERVE,
                ),
                NodeInsert(
                    row_id=new_row_id(),
                    logical_id="adaptive-goal-roadmap",
                    kind=_ADAPTIVE_GOAL_KIND,
                    properties={"title": "Engineering roadmap"},
                    source_ref="adaptive-seed",
                    upsert=False,
                    chunk_policy=ChunkPolicy.PRESERVE,
                ),
            ],
            chunks=[
                ChunkInsert(
                    id="adaptive-goal-budget-chunk",
                    node_logical_id="adaptive-goal-budget",
                    text_content="budget meeting notes for finance review",
                ),
                ChunkInsert(
                    id="adaptive-goal-quarterly-chunk",
                    node_logical_id="adaptive-goal-quarterly",
                    text_content="quarterly planning docs and action items",
                ),
                ChunkInsert(
                    id="adaptive-goal-roadmap-chunk",
                    node_logical_id="adaptive-goal-roadmap",
                    text_content="engineering roadmap deliverables",
                ),
            ],
        )
    )


def adaptive_search_strict_hit_only(context: HarnessContext) -> ScenarioResult:
    """Strict branch finds hits — relaxed fallback never fires."""
    db_path = context.sibling_db("adaptive-strict-hit-only")
    engine = context.open_engine(db_path)
    try:
        _seed_adaptive_goals(engine)
        rows = (
            engine.nodes(_ADAPTIVE_GOAL_KIND)
            .text_search("budget meeting", 10)
            .execute()
        )
        assert len(rows.hits) == 1, f"expected 1 hit, got {len(rows.hits)}"
        hit = rows.hits[0]
        assert hit.node.logical_id == "adaptive-goal-budget", (
            f"unexpected hit {hit.node.logical_id!r}"
        )
        assert hit.match_mode == SearchMatchMode.STRICT, (
            f"expected STRICT, got {hit.match_mode}"
        )
        assert rows.fallback_used is False, "fallback must not fire on strict hit"
        assert rows.was_degraded is False, "search must not degrade"
        assert rows.strict_hit_count == 1, (
            f"strict_hit_count={rows.strict_hit_count}"
        )
        assert rows.relaxed_hit_count == 0, (
            f"relaxed_hit_count={rows.relaxed_hit_count}"
        )
    finally:
        engine.close()
    return ScenarioResult(name="adaptive_search_strict_hit_only")


def adaptive_search_strict_miss_relaxed_recovery(
    context: HarnessContext,
) -> ScenarioResult:
    """Strict branch misses on an implicit-AND term; relaxed branch recovers."""
    db_path = context.sibling_db("adaptive-relaxed-recovery")
    engine = context.open_engine(db_path)
    try:
        _seed_adaptive_goals(engine)
        rows = (
            engine.nodes(_ADAPTIVE_GOAL_KIND)
            .text_search("budget nonexistentxyzzy", 10)
            .execute()
        )
        assert rows.fallback_used is True, (
            "strict miss must trigger the relaxed fallback branch"
        )
        assert len(rows.hits) > 0, "relaxed branch should recover at least one hit"
        assert rows.relaxed_hit_count > 0, (
            f"relaxed_hit_count={rows.relaxed_hit_count}"
        )
        assert any(h.match_mode == SearchMatchMode.RELAXED for h in rows.hits), (
            "at least one hit must carry match_mode=RELAXED"
        )
    finally:
        engine.close()
    return ScenarioResult(name="adaptive_search_strict_miss_relaxed_recovery")


def adaptive_search_mixed_chunk_and_property(
    context: HarnessContext,
) -> ScenarioResult:
    """Collapse chunk+property hits for the same node into a single chunk hit."""
    db_path = context.sibling_db("adaptive-mixed-chunk-property")
    engine = context.open_engine(db_path)
    try:
        # Register a scalar property schema on $.title so the property
        # branch matches the same term that the chunk branch matches.
        engine.admin.register_fts_property_schema(_ADAPTIVE_GOAL_KIND, ["$.title"])
        engine.write(
            WriteRequest(
                label="adaptive-dual-match",
                nodes=[
                    NodeInsert(
                        row_id=new_row_id(),
                        logical_id="adaptive-dual",
                        kind=_ADAPTIVE_GOAL_KIND,
                        properties={"title": "dualmatchneedle target"},
                        source_ref="adaptive-seed",
                        upsert=False,
                        chunk_policy=ChunkPolicy.PRESERVE,
                    )
                ],
                chunks=[
                    ChunkInsert(
                        id="adaptive-dual-chunk",
                        node_logical_id="adaptive-dual",
                        text_content="dualmatchneedle target appears in chunk body",
                    )
                ],
            )
        )
        rows = (
            engine.nodes(_ADAPTIVE_GOAL_KIND)
            .text_search("dualmatchneedle", 10)
            .execute()
        )
        assert len(rows.hits) == 1, (
            f"dedup must collapse chunk+property to one hit; got {len(rows.hits)}"
        )
        hit = rows.hits[0]
        assert hit.node.logical_id == "adaptive-dual", (
            f"unexpected hit {hit.node.logical_id!r}"
        )
        assert hit.source == SearchHitSource.CHUNK, (
            f"chunk must win the source tiebreak; got {hit.source}"
        )
    finally:
        engine.close()
    return ScenarioResult(name="adaptive_search_mixed_chunk_and_property")


def _register_recursive_payload_schema(engine) -> None:
    engine.admin.register_fts_property_schema_with_entries(
        _ADAPTIVE_KI_KIND,
        [FtsPropertyPathSpec(path="$.payload", mode=FtsPropertyPathMode.RECURSIVE)],
        separator=" ",
        exclude_paths=[],
    )


def _seed_recursive_knowledge_item(engine) -> None:
    engine.write(
        WriteRequest(
            label="adaptive-seed-ki",
            nodes=[
                NodeInsert(
                    row_id=new_row_id(),
                    logical_id="adaptive-ki-alpha",
                    kind=_ADAPTIVE_KI_KIND,
                    properties={
                        "payload": {
                            "title": "quarterly planning",
                            "notes": "budget approval",
                        }
                    },
                    source_ref="adaptive-seed",
                    upsert=False,
                    chunk_policy=ChunkPolicy.PRESERVE,
                )
            ],
        )
    )


def adaptive_search_recursive_nested_payload(
    context: HarnessContext,
) -> ScenarioResult:
    """Recursive-property schema populates attribution matched_paths."""
    db_path = context.sibling_db("adaptive-recursive-payload")
    engine = context.open_engine(db_path)
    try:
        _register_recursive_payload_schema(engine)
        _seed_recursive_knowledge_item(engine)

        rows = (
            engine.nodes(_ADAPTIVE_KI_KIND)
            .text_search("quarterly", 10)
            .with_match_attribution()
            .execute()
        )
        assert len(rows.hits) == 1, f"expected 1 hit, got {len(rows.hits)}"
        hit = rows.hits[0]
        assert hit.source == SearchHitSource.PROPERTY, (
            f"recursive property hit must report PROPERTY source; got {hit.source}"
        )
        assert hit.match_mode == SearchMatchMode.STRICT, (
            f"expected STRICT, got {hit.match_mode}"
        )
        assert hit.attribution is not None, "attribution must be populated"
        assert "$.payload.title" in hit.attribution.matched_paths, (
            f"expected $.payload.title in matched_paths; got "
            f"{hit.attribution.matched_paths!r}"
        )

        # Multi-term: both leaves must appear in matched_paths in
        # first-match-offset order.
        rows_multi = (
            engine.nodes(_ADAPTIVE_KI_KIND)
            .text_search("quarterly budget", 10)
            .with_match_attribution()
            .execute()
        )
        assert len(rows_multi.hits) == 1, (
            f"expected 1 hit for multi-term, got {len(rows_multi.hits)}"
        )
        multi_hit = rows_multi.hits[0]
        assert multi_hit.attribution is not None, (
            "attribution must be populated on multi-term hit"
        )
        paths = multi_hit.attribution.matched_paths
        assert "$.payload.title" in paths, (
            f"expected $.payload.title in matched_paths; got {paths!r}"
        )
        assert "$.payload.notes" in paths, (
            f"expected $.payload.notes in matched_paths; got {paths!r}"
        )
    finally:
        engine.close()
    return ScenarioResult(name="adaptive_search_recursive_nested_payload")


def adaptive_search_recursive_rebuild_restore(
    context: HarnessContext,
) -> ScenarioResult:
    """Recursive schema + position map survive a close/re-open round-trip."""
    db_path = context.sibling_db("adaptive-recursive-reopen")
    engine = context.open_engine(db_path)
    try:
        _register_recursive_payload_schema(engine)
        _seed_recursive_knowledge_item(engine)

        rows_before = (
            engine.nodes(_ADAPTIVE_KI_KIND)
            .text_search("quarterly", 10)
            .with_match_attribution()
            .execute()
        )
        assert len(rows_before.hits) == 1, (
            f"pre-reopen expected 1 hit, got {len(rows_before.hits)}"
        )
        hit_before = rows_before.hits[0]
        assert hit_before.attribution is not None
        paths_before = hit_before.attribution.matched_paths
    finally:
        engine.close()

    engine = context.open_engine(db_path)
    try:
        rows_after = (
            engine.nodes(_ADAPTIVE_KI_KIND)
            .text_search("quarterly", 10)
            .with_match_attribution()
            .execute()
        )
        assert len(rows_after.hits) == 1, (
            f"post-reopen expected 1 hit, got {len(rows_after.hits)}"
        )
        hit_after = rows_after.hits[0]
        assert hit_after.node.logical_id == hit_before.node.logical_id, (
            "logical_id must match across reopen"
        )
        assert hit_after.source == hit_before.source, (
            f"source must match across reopen; before={hit_before.source}, "
            f"after={hit_after.source}"
        )
        assert hit_after.attribution is not None, (
            "attribution must still populate after reopen"
        )
        assert hit_after.attribution.matched_paths == paths_before, (
            f"matched_paths must match across reopen; before={paths_before!r}, "
            f"after={hit_after.attribution.matched_paths!r}"
        )
    finally:
        engine.close()
    return ScenarioResult(name="adaptive_search_recursive_rebuild_restore")

File: archive/pre-0.6.0/typescript/apps/sdk-harness/src/scenarios/adaptive-search.ts
import {
  Engine,
  WriteRequestBuilder,
  newRowId,
  type SearchRows,
} from "fathomdb";
import type { HarnessResult } from "../models.js";
import { assert, handleScenarioError, tempDbPath } from "../skip.js";

// Dedicated kinds so these scenarios do not collide with other
// harness runs that also write to a tempdir-backed engine.
const GOAL_KIND = "AdaptiveGoal";
const KI_KIND = "AdaptiveKnowledgeItem";

function seedAdaptiveGoals(engine: Engine): void {
  const builder = new WriteRequestBuilder("adaptive-seed-goals");
  const budget = builder.addNode({
    rowId: newRowId(),
    logicalId: "adaptive-goal-budget",
    kind: GOAL_KIND,
    properties: { title: "Budget meeting" },
    sourceRef: "adaptive-seed",
    upsert: false,
    chunkPolicy: "preserve",
  });
  const quarterly = builder.addNode({
    rowId: newRowId(),
    logicalId: "adaptive-goal-quarterly",
    kind: GOAL_KIND,
    properties: { title: "Quarterly planning" },
    sourceRef: "adaptive-seed",
    upsert: false,
    chunkPolicy: "preserve",
  });
  const roadmap = builder.addNode({
    rowId: newRowId(),
    logicalId: "adaptive-goal-roadmap",
    kind: GOAL_KIND,
    properties: { title: "Engineering roadmap" },
    sourceRef: "adaptive-seed",
    upsert: false,
    chunkPolicy: "preserve",
  });
  builder.addChunk({
    id: "adaptive-goal-budget-chunk",
    node: budget,
    textContent: "budget meeting notes for finance review",
  });
  builder.addChunk({
    id: "adaptive-goal-quarterly-chunk",
    node: quarterly,
    textContent: "quarterly planning docs and action items",
  });
  builder.addChunk({
    id: "adaptive-goal-roadmap-chunk",
    node: roadmap,
    textContent: "engineering roadmap deliverables",
  });
  engine.write(builder.build());
}

function registerRecursivePayloadSchema(engine: Engine): void {
  engine.admin.registerFtsPropertySchemaWithEntries({
    kind: KI_KIND,
    entries: [{ path: "$.payload", mode: "recursive" }],
    separator: " ",
    excludePaths: [],
  });
}

function seedRecursiveKnowledgeItem(engine: Engine): void {
  const builder = new WriteRequestBuilder("adaptive-seed-ki");
  builder.addNode({
    rowId: newRowId(),
    logicalId: "adaptive-ki-alpha",
    kind: KI_KIND,
    properties: {
      payload: {
        title: "quarterly planning",
        notes: "budget approval",
      },
    },
    sourceRef: "adaptive-seed",
    upsert: false,
    chunkPolicy: "preserve",
  });
  engine.write(builder.build());
}

function strictHitOnly(): HarnessResult {
  try {
    const engine = Engine.open(tempDbPath("adaptive-strict-hit-only"));
    try {
      seedAdaptiveGoals(engine);
      const rows: SearchRows = engine
        .query(GOAL_KIND)
        .textSearch("budget meeting", 10)
        .execute();
      assert(rows.hits.length === 1, `expected 1 hit, got ${rows.hits.length}`);
      const hit = rows.hits[0];
      assert(
        hit.node.logicalId === "adaptive-goal-budget",
        `unexpected hit logicalId=${hit.node.logicalId}`,
      );
      assert(hit.matchMode === "strict", `expected matchMode=strict, got ${hit.matchMode}`);
      assert(rows.fallbackUsed === false, "fallback must not fire on strict hit");
      assert(rows.wasDegraded === false, "search must not degrade");
      assert(rows.strictHitCount === 1, `strictHitCount=${rows.strictHitCount}`);
      assert(rows.relaxedHitCount === 0, `relaxedHitCount=${rows.relaxedHitCount}`);
    } finally {
      engine.close();
    }
    return { name: "adaptive_search_strict_hit_only", ok: true };
  } catch (error) {
    return handleScenarioError("adaptive_search_strict_hit_only", error);
  }
}

function strictMissRelaxedRecovery(): HarnessResult {
  try {
    const engine = Engine.open(tempDbPath("adaptive-relaxed-recovery"));
    try {
      seedAdaptiveGoals(engine);
      const rows = engine
        .query(GOAL_KIND)
        .textSearch("budget nonexistentxyzzy", 10)
        .execute();
      assert(rows.fallbackUsed === true, "strict miss must trigger relaxed fallback");
      assert(rows.hits.length > 0, "relaxed branch must recover at least one hit");
      assert(
        rows.relaxedHitCount > 0,
        `relaxedHitCount=${rows.relaxedHitCount}`,
      );
      assert(
        rows.hits.some((h) => h.matchMode === "relaxed"),
        "at least one hit must carry matchMode=relaxed",
      );
    } finally {
      engine.close();
    }
    return { name: "adaptive_search_strict_miss_relaxed_recovery", ok: true };
  } catch (error) {
    return handleScenarioError("adaptive_search_strict_miss_relaxed_recovery", error);
  }
}

function mixedChunkAndProperty(): HarnessResult {
  try {
    const engine = Engine.open(tempDbPath("adaptive-mixed-chunk-property"));
    try {
      engine.admin.registerFtsPropertySchema(GOAL_KIND, ["$.title"]);
      const builder = new WriteRequestBuilder("adaptive-dual-match");
      const dual = builder.addNode({
        rowId: newRowId(),
        logicalId: "adaptive-dual",
        kind: GOAL_KIND,
        properties: { title: "dualmatchneedle target" },
        sourceRef: "adaptive-seed",
        upsert: false,
        chunkPolicy: "preserve",
      });
      builder.addChunk({
        id: "adaptive-dual-chunk",
        node: dual,
        textContent: "dualmatchneedle target appears in chunk body",
      });
      engine.write(builder.build());

      const rows = engine
        .query(GOAL_KIND)
        .textSearch("dualmatchneedle", 10)
        .execute();
      assert(
        rows.hits.length === 1,
        `dedup must collapse chunk+property to one hit; got ${rows.hits.length}`,
      );
      const hit = rows.hits[0];
      assert(
        hit.node.logicalId === "adaptive-dual",
        `unexpected hit logicalId=${hit.node.logicalId}`,
      );
      assert(
        hit.source === "chunk",
        `chunk must win the source tiebreak; got source=${hit.source}`,
      );
    } finally {
      engine.close();
    }
    return { name: "adaptive_search_mixed_chunk_and_property", ok: true };
  } catch (error) {
    return handleScenarioError("adaptive_search_mixed_chunk_and_property", error);
  }
}

function recursiveNestedPayload(): HarnessResult {
  try {
    const engine = Engine.open(tempDbPath("adaptive-recursive-payload"));
    try {
      registerRecursivePayloadSchema(engine);
      seedRecursiveKnowledgeItem(engine);

      const rows = engine
        .query(KI_KIND)
        .textSearch("quarterly", 10)
        .withMatchAttribution()
        .execute();
      assert(rows.hits.length === 1, `expected 1 hit, got ${rows.hits.length}`);
      const hit = rows.hits[0];
      assert(
        hit.source === "property",
        `recursive property hit must report source=property; got ${hit.source}`,
      );
      assert(
        hit.matchMode === "strict",
        `expected matchMode=strict, got ${hit.matchMode}`,
      );
      assert(hit.attribution !== null, "attribution must be populated");
      const paths = hit.attribution!.matchedPaths;
      assert(
        paths.includes("$.payload.title"),
        `expected $.payload.title in matchedPaths; got ${JSON.stringify(paths)}`,
      );

      const multi = engine
        .query(KI_KIND)
        .textSearch("quarterly budget", 10)
        .withMatchAttribution()
        .execute();
      assert(
        multi.hits.length === 1,
        `expected 1 hit on multi-term, got ${multi.hits.length}`,
      );
      const multiHit = multi.hits[0];
      assert(multiHit.attribution !== null, "multi-term attribution must be populated");
      const multiPaths = multiHit.attribution!.matchedPaths;
      assert(
        multiPaths.includes("$.payload.title"),
        `expected $.payload.title; got ${JSON.stringify(multiPaths)}`,
      );
      assert(
        multiPaths.includes("$.payload.notes"),
        `expected $.payload.notes; got ${JSON.stringify(multiPaths)}`,
      );
    } finally {
      engine.close();
    }
    return { name: "adaptive_search_recursive_nested_payload", ok: true };
  } catch (error) {
    return handleScenarioError("adaptive_search_recursive_nested_payload", error);
  }
}

function recursiveRebuildRestore(): HarnessResult {
  try {
    const dbPath = tempDbPath("adaptive-recursive-reopen");
    let pathsBefore: string[] = [];
    let logicalIdBefore = "";
    let sourceBefore = "";

    {
      const engine = Engine.open(dbPath);
      try {
        registerRecursivePayloadSchema(engine);
        seedRecursiveKnowledgeItem(engine);
        const rowsBefore = engine
          .query(KI_KIND)
          .textSearch("quarterly", 10)
          .withMatchAttribution()
          .execute();
        assert(
          rowsBefore.hits.length === 1,
          `pre-reopen expected 1 hit, got ${rowsBefore.hits.length}`,
        );
        const hit = rowsBefore.hits[0];
        assert(hit.attribution !== null, "pre-reopen attribution must be populated");
        pathsBefore = [...hit.attribution!.matchedPaths];
        logicalIdBefore = hit.node.logicalId;
        sourceBefore = hit.source;
      } finally {
        engine.close();
      }
    }

    {
      const engine = Engine.open(dbPath);
      try {
        const rowsAfter = engine
          .query(KI_KIND)
          .textSearch("quarterly", 10)
          .withMatchAttribution()
          .execute();
        assert(
          rowsAfter.hits.length === 1,
          `post-reopen expected 1 hit, got ${rowsAfter.hits.length}`,
        );
        const hit = rowsAfter.hits[0];
        assert(
          hit.node.logicalId === logicalIdBefore,
          `logicalId mismatch across reopen: before=${logicalIdBefore}, after=${hit.node.logicalId}`,
        );
        assert(
          hit.source === sourceBefore,
          `source mismatch across reopen: before=${sourceBefore}, after=${hit.source}`,
        );
        assert(
          hit.attribution !== null,
          "post-reopen attribution must be populated",
        );
        const pathsAfter = hit.attribution!.matchedPaths;
        assert(
          pathsAfter.length === pathsBefore.length &&
            pathsAfter.every((p, i) => p === pathsBefore[i]),
          `matchedPaths mismatch across reopen: before=${JSON.stringify(pathsBefore)}, after=${JSON.stringify(pathsAfter)}`,
        );
      } finally {
        engine.close();
      }
    }
    return { name: "adaptive_search_recursive_rebuild_restore", ok: true };
  } catch (error) {
    return handleScenarioError("adaptive_search_recursive_rebuild_restore", error);
  }
}

export function adaptiveSearchScenarios(): HarnessResult[] {
  return [
    strictHitOnly(),
    strictMissRelaxedRecovery(),
    mixedChunkAndProperty(),
    recursiveNestedPayload(),
    recursiveRebuildRestore(),
  ];
}
